tycho-p2-repository-plugin/src/main/java/org/eclipse/tycho/plugins/p2/repository/AssembleRepositoryMojo.java

```diff
@@ -78,8 +74,19 @@
 public class AssembleRepositoryMojo extends AbstractRepositoryMojo {
 
     public static class RepositoryReferenceFilter {
-        List<String> exclude;
-        List<String> include;
+        /**
+         * If repository references providing no relevant content or only content also provided by
+         * others, should be excluded from being added automatically as referenced repositories of
+         * the assembled repository.
+         */
+        boolean excludeNotStrictlyRequired = true;
+        /** The list of location patterns that exclude matching repository references. */
+        List<String> exclude = List.of();
+        /**
+         * The list of location patterns that includes only matching repository references (if this
+         * list is not empty).
+         */
+        List<String> include = List.of();
```

Comment:
I think this must be made more clear, what happens if I exclude `A*` but include `A/B/*` will `A/B/C` contained in the repository? Or can/should I always only choose either `exclude` or `include`?

Reply:
Maybe `<matches>` instead of exclude would be better (the user already knows we are in a filtering expression so exclusion is implicit) would be better

Reply:
> I think this must be made more clear, what happens if I exclude `A*` but include `A/B/*` will `A/B/C` contained in the repository? Or can/should I always only choose either `exclude` or `include`?

You can choose both.
I have described that in the doc of the `repositoryReferenceFilter` parameter.
The inclusions are applied first, if any are specified, which means among all available references it is first checked which match the inclusion filter and all others are discarded (if no inclusion is specified just all available refs are considered for the next step). On the remaining references in step two all reference that match an exclusion are discarded too.
So basically the inclusion is just an negated exclusion.

Therefore in the case you described A/B/C will not be contained in the assembled repo.

Reply:
> The inclusions are applied first, if any are specified, which means among all available references it is first checked which match the inclusion filter and all others are discarded (if no inclusion is specified just all available refs are considered for the next step). On the remaining references in step two all reference that match an exclusion are discarded too.

Hm... I would expect it the other way round as by default everything is included already, so maybe `matches` is a better wording here?

Reply:
> Hm... I would expect it the other way round as by default everything is included already, so maybe `matches` is a better wording here?

It probably quite personal what one expects. On the other hand it would be somehow strange as well to eventually include things that matched a exclusion?

But since inclusion and exclusion basically just have an inverted sign, what about adding a syntax to the pattern to invert it and just have exclusions/excluding filter? E.g. similar to the style `%regex[<pattern>]` we could for example use `![<pattern>]` to negate a pattern.
And instead of 
```
<repositoryReferenceFilter>
    <exclude>https://foo.bar/hidden</exclude>
    <include>https://foo.bar</include>
<repositoryReferenceFilter>
```
we would only have something like
```
<repositoryReferenceFilter>
    <exclude>https://foo.bar/hidden</exclude>
    <exclude>![https://foo.bar]</exclude>
<repositoryReferenceFilter>
```
Plus the parameter for the new automated removal of unused references.

Btw. regarding expecting exclusion during filtering: E.g. in the Java-Stream API the `filter()` method expects a Predicate which all elements must match in order to pass the filter.
So I think it is not so clear that filter means always exclusion (although I would also tend more towards exclusion).

Reply:
> It probably quite personal what one expects. On the other hand it would be somehow strange as well to eventually include things that matched a exclusion?

The problem often is I want to block everything except one, for example in `.gitignore` it could also be quite challenging:
https://stackoverflow.com/questions/5533050/gitignore-exclude-folder-but-include-specific-subfolder

> Btw. regarding expecting exclusion during filtering: E.g. in the Java-Stream API the filter() method expects a Predicate which all elements must match in order to pass the filter.

Yes that's very confusing and hit me several times already :-\

Reply:
I think this is now solved with https://github.com/eclipse-tycho/tycho/pull/2849.